components/brave_wallet/browser/cardano/cardano_knapsack_solver.cc

```diff
@@ -113,40 +69,20 @@ void CardanoKnapsackSolver::RunSolverForTransaction(
         CardanoTransaction next_transaction = cur_transaction;
         next_transaction.AddInput(inputs_[input_index]);
 
-        if (next_transaction.ChangeOutput()) {
-          // Calculate tx fee and move everything surplus to change output.
-          // Throw away possible transaction if fee requirements could no be
-          // satisfied.
-          if (!SetFeeAndChangeForTransaction(next_transaction,
-                                             latest_epoch_parameters_)) {
-            continue;
-          }
-
-          // Throw away possible transaction if resulting change amount is
-          // less than min ADA for output threshold.
-          if (!CardanoTransactionSerializer().ValidateMinValue(
-                  *next_transaction.ChangeOutput(), latest_epoch_parameters_)) {
-            continue;
-          }
-        }
-
-        // Minimum fee required for this transaction to be accepted.
-        // Depends on transaction's size and current fee rate.
-        uint64_t min_fee =
-            CardanoTransactionSerializer({.use_dummy_witness_set = true})
-                .CalcMinTransactionFee(next_transaction,
-                                       latest_epoch_parameters_);
-
-        if (next_transaction.AmountsAreValid(min_fee)) {
+        // Given set of inputs and outputs try to find valid outputs' amounts
+        // and fee. If it fails to find a valid transaction: proceed to the next
+        // iteration with current input. Otherwise check if it is the best
+        // tx so far and proceed to the next iteration without current input.
+        if (!CardanoTransactionSerializer::AdjustFeeAndOutputsForTx(
+                next_transaction, latest_epoch_parameters_)) {
+          picked_inputs[input_index] = true;
+          cur_transaction = std::move(next_transaction);
+        } else {
           has_valid_transaction_for_iteration = true;
           if (!current_best_solution_ ||
-              current_best_solution_->EffectiveFeeAmount() >
-                  next_transaction.EffectiveFeeAmount()) {
+              current_best_solution_->fee() > next_transaction.fee()) {
```

Comment:
In some rare cases it seems next_transaction could be in incorrect state if AdjustFeeAndOutputsForTx fails, for example change could be 0